src/android/usb_host/android_uvc.cpp

```diff
@@ -0,0 +1,1804 @@
+                    cur_format->fourcc = SWAP_UINT32(*(const uint32_t *) format->guidFormat);
+                    if(1496850464 == cur_format->fourcc)
+                        cur_format->fourcc = 1196574041; //TODO
+                    cur_format->interfaceNumber = stream_if->bInterfaceNumber;
+
+                    cur_format->fps = 10000000 / *interval_ptr;
+                    cur_format->next = NULL;
+
+                    if (prev_format != NULL) {
+                        prev_format->next = cur_format;
+                    } else {
+                        *formats = cur_format;
+                    }
+
+                    prev_format = cur_format;
+                }
+            }
+        }
+    }
+    return UVC_SUCCESS;
+}
+
+uvc_error_t usbhost_free_formats(uvc_format_t *formats) {
+    uvc_format_t *cur_format = formats;
+    while (cur_format != NULL) {
+        uvc_format_t *format = cur_format;
+        cur_format = cur_format->next;
+        free(format);
+    }
+
+    return UVC_SUCCESS;
+}
+
+/** @internal
+* @brief Find the descriptor for a specific frame configuration
+* @param stream_if Stream interface
+* @param format_id Index of format class descriptor
+* @param frame_id Index of frame descriptor
+*/
+static uvc_frame_desc_t *
+usbhost_uvc_find_frame_desc_stream_if(usbhost_uvc_streaming_interface_t *stream_if,
+                                      uint16_t format_id, uint16_t frame_id) {
+
+    uvc_format_desc_t *format = NULL;
+    uvc_frame_desc_t *frame = NULL;
+
+    DL_FOREACH(stream_if->format_descs, format) {
+        if (format->bFormatIndex == format_id) {
+            DL_FOREACH(format->frame_descs, frame) {
+                if (frame->bFrameIndex == frame_id)
+                    return frame;
+            }
+        }
+    }
+
+    return NULL;
+}
+
+uvc_frame_desc_t *usbhost_uvc_find_frame_desc_stream(usbhost_uvc_stream_handle_t *strmh,
+                                                     uint16_t format_id, uint16_t frame_id) {
+    return usbhost_uvc_find_frame_desc_stream_if(strmh->stream_if, format_id, frame_id);
+}
+
+void usbhost_uvc_process_payload(usbhost_uvc_stream_handle_t* stream_handle, frames_archive *archive, frames_queue *queue) {
+
+    const size_t payload_length = stream_handle->got_bytes;
+
+    // Nothing to do for empty requests.
+    if (payload_length == 0) {
+        return;
+    }
+
+    const uint8_t header_length = stream_handle->outbuf[0];
+
+    // Check if the payload length is greater than the received bytes.
+    if (header_length > payload_length) {
+        LOG_ERROR("bogus packet: payload_length=" << payload_length << ", header_length=" << (int)header_length);
+        return;
+    }
+
+    // Check standard UVC header length.
+    if (header_length != 12) {
+        LOG_ERROR("bogus header: header_length=" << (int)header_length);
+    }
+
+    uint8_t header_info = 0;
+
+    //12 bytes - standard UVC header
+    if (header_length == 12) {
```

Comment:
This if statement and the one above it could be combined in to a single if-else. Also I expect a header that isn't 12 bytes long should probably be an early return as well.

Reply:
Done in 4830dc2b3